builtInFunctions/esdtDeleteMetadata.go

```diff
@@ -0,0 +1,314 @@
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		err := e.addMetadata(vmInput.Arguments)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	vmOutput := &vmcommon.VMOutput{ReturnCode: vmcommon.Ok}
+
+	return vmOutput, nil
+}
+
+// input is list(tokenID-numIntervals-list(start,end))
+func (e *esdtDeleteMetaData) deleteMetadata(args [][]byte) error {
+	lenArgs := uint64(len(args))
+	if lenArgs < 4 {
+		return ErrInvalidNumOfArgs
+	}
+
+	systemAcc, err := e.getSystemAccount()
+	if err != nil {
+		return err
+	}
+
+	for i := uint64(0); i+1 < uint64(len(args)); {
+		tokenID := args[i]
+		numIntervals := big.NewInt(0).SetBytes(args[i+1]).Uint64()
+		i += 2
+
+		if !vmcommon.ValidateToken(tokenID) {
+			return ErrInvalidTokenID
+		}
+
+		if i >= lenArgs {
+			return ErrInvalidNumOfArgs
+		}
+
+		err = e.deleteMetadataForListIntervals(systemAcc, tokenID, args, i, numIntervals)
+		if err != nil {
+			return err
+		}
+
+		i += numIntervals * 2
+	}
+
+	err = e.accounts.SaveAccount(systemAcc)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (e *esdtDeleteMetaData) deleteMetadataForListIntervals(
+	systemAcc vmcommon.UserAccountHandler,
+	tokenID []byte,
+	args [][]byte,
+	index, numIntervals uint64,
+) error {
+	lenArgs := uint64(len(args))
+	for j := index; j < index+numIntervals*2; j += 2 {
+		if j > lenArgs-2 {
+			return ErrInvalidNumOfArgs
+		}
+
+		startIndex := big.NewInt(0).SetBytes(args[j]).Uint64()
+		endIndex := big.NewInt(0).SetBytes(args[j+1]).Uint64()
+
+		err := e.deleteMetadataForInterval(systemAcc, tokenID, startIndex, endIndex)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (e *esdtDeleteMetaData) deleteMetadataForInterval(
+	systemAcc vmcommon.UserAccountHandler,
+	tokenID []byte,
+	startIndex, endIndex uint64,
+) error {
+	if endIndex < startIndex {
+		return ErrInvalidArguments
+	}
+	if startIndex == 0 {
+		return ErrInvalidNonce
+	}
+
+	esdtTokenKey := append(e.keyPrefix, tokenID...)
+	for nonce := startIndex; nonce <= endIndex; nonce++ {
+		esdtNFTTokenKey := computeESDTNFTTokenKey(esdtTokenKey, nonce)
+
+		err := systemAcc.AccountDataHandler().SaveKeyValue(esdtNFTTokenKey, nil)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// input is list(tokenID-nonce-metadata)
+func (e *esdtDeleteMetaData) addMetadata(args [][]byte) error {
+	if len(args)%numArgsPerAdd != 0 || len(args) < numArgsPerAdd {
+		return ErrInvalidNumOfArgs
+	}
+
+	systemAcc, err := e.getSystemAccount()
+	if err != nil {
+		return err
+	}
+
+	for i := 0; i < len(args); i += numArgsPerAdd {
+		tokenID := args[i]
+		nonce := big.NewInt(0).SetBytes(args[i+1]).Uint64()
+		if nonce == 0 {
+			return ErrInvalidNonce
+		}
+
+		if !vmcommon.ValidateToken(tokenID) {
+			return ErrInvalidTokenID
+		}
+
+		esdtTokenKey := append(e.keyPrefix, tokenID...)
+		esdtNFTTokenKey := computeESDTNFTTokenKey(esdtTokenKey, nonce)
+		metaData := &esdt.MetaData{}
+		err = e.marshalizer.Unmarshal(metaData, args[i+2])
```

Comment:
might also add some checks for the unmarshalled metaData (ex: nonce should be higher than 0). Otherwise, allowed addresses will be able to store anything that could make no sense to the protocol

Reply:
done